components/lsp-form/MintForm.tsx

```diff
@@ -9,19 +9,109 @@ import {
+  lspContract: ethers.Contract | null;
+  erc20Contract: ethers.Contract | null;
+  web3Provider: ethers.providers.Web3Provider | null;
+  collateralBalance: ethers.BigNumber;
+  tokensMinted: ethers.BigNumber;
+  collateralPerPair: string;
+  refetchTokensCreatedEvents: (
+    options?: RefetchOptions | undefined
+  ) => Promise<
+    QueryObserverResult<void | TokensCreated[] | undefined, unknown>
+  >;
+  setCollateralBalance: React.Dispatch<React.SetStateAction<ethers.BigNumber>>;
+  erc20Decimals: string;
+  collateralDecimals: string;
 }
-const MintForm: FC<Props> = ({ setShowSettle }) => {
-  const [collateral, setCollateral] = useState("");
+
+const MintForm: FC<Props> = ({
+  // setShowSettle,
+  lspContract,
+  erc20Contract,
+  contractAddress,
+  collateralBalance,
+  tokensMinted,
+  collateralPerPair,
+  refetchTokensCreatedEvents,
+  address,
+  setCollateralBalance,
+  erc20Decimals,
+  collateralDecimals,
+}) => {
+  const [collateral, setCollateral] = useState("uma");
   const [amount, setAmount] = useState("");
   const [longTokenAmount, setLongTokenAmount] = useState("");
   const [shortTokenAmount, setShortTokenAmount] = useState("");
 
+  const mint = useCallback(async () => {
+    if (lspContract && erc20Contract && amount) {
+      const weiAmount = toWeiSafe(amount);
+      try {
+        const allowance = await erc20Contract.allowance(
+          address,
+          contractAddress
+        );
+        const balance = await erc20Contract.balanceOf(address);
+        const hasToApprove = allowance.lt(balance);
+        if (hasToApprove) {
+          const approveTx = await erc20Contract.approve(
+            contractAddress,
+            INFINITE_APPROVAL_AMOUNT
+          );
+
+          await approveTx.wait();
+        }
+        // Need to send the correct amount based on the collateral pair ** the amount
+        // User has specified in the input.
+        // Get rid of float, then divide it out later.
+        const ratio = ethers.BigNumber.from(1 / Number(collateralPerPair)).mul(
+          ethers.BigNumber.from(10).pow(18)
+        );
+
+        lspContract
+          .create(weiAmount.mul(ratio).div(ethers.BigNumber.from(10).pow(18)))
+          .then((tx: any) => {
+            tx.wait(1).then(async () => {
+              setAmount("");
+              setLongTokenAmount("");
+              setShortTokenAmount("");
+              refetchTokensCreatedEvents();
+              const balance = (await erc20Contract.balanceOf(
+                address
+              )) as ethers.BigNumber;
+              setCollateralBalance(balance);
+            });
+          });
+      } catch (err) {
+        console.log("err", err);
+      }
+    }
+  }, [lspContract, erc20Contract, amount]);
+
+  useEffect(() => {
+    if (amount !== "") {
+      const tokenAmounts = Number(amount) / Number(collateralPerPair);
+      setLongTokenAmount(tokenAmounts.toString());
+      setShortTokenAmount(tokenAmounts.toString());
```

Comment:
i still think thers a potential problem here, particularly with very small floats. why not allow a side effect on the setAmount input box, and have it setShort/setLong tokens directly?  so now you have a directional channel where when user is inputting in long/short inputs, that will setAmount and when user editing amount input, it sets long/short token.  you could also just combine long/short token amount into a single number since they should always be the same

Reply:
Yeah totally reasonable, I can use the same flow as the other 2 and avoid useEffect altogether. It also self contains the update close to the actual input which is actually nice.